### PR TITLE
Task 86737 Refactor CLI for IOC

### PR DESCRIPTION
**Pull request by nikhilgirraj**
Added interfaces for IApiConsumer, IConsoleSubscriberWriter, ISubscribersDirectoryProcessor. 
Registered dependencies.
Added more tests.

### CODE REVIEW

File: src/Platform.Eda.Cli/Commands/ConfigureEda/ConfigureEdaCommand.cs [hunk 1]
```diff
@@ -51,19 +51,16 @@ public ConfigureEdaCommand(IFileSystem fileSystem, ICaptainHookClient captainHoo
             ShowInHelpText = true)]
         public bool NoDryRun { get; set; }
 
-        public async Task<int> OnExecuteAsync(CommandLineApplication app, IConsole console)
+        public async Task<int> OnExecuteAsync(CommandLineApplication app, IConsole console, IConsoleSubscriberWriter writer)
```
Comment (mszymeshop) [inline src/Platform.Eda.Cli/Commands/ConfigureEda/ConfigureEdaCommand.cs:54]: is `console` used in that case at all?

File: src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+        private readonly ConfigureEdaCommand _configureEdaCommand;
+        private readonly Mock<ICaptainHookClient> _mockCaptainHookClient;
+        private readonly IApiConsumer _apiConsumer;
+        private readonly IConsoleSubscriberWriter _mockConsoleSubscriberWriter;
+
+        public ConfigureEdaCommandTests(ITestOutputHelper output) : base(output)
+        {
+            _mockCaptainHookClient = new Mock<ICaptainHookClient>();
+            _mockConsoleSubscriberWriter = new ConsoleSubscriberWriter(Console);
+            _apiConsumer = new ApiConsumer(_mockCaptainHookClient.Object, null);
```
Comment (mszymeshop) [inline src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs:34]: You should mock `ApiConsumer` for the purpose of Command tests

File: src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+
+            var subscribersDirectoryProcessor = new SubscribersDirectoryProcessor(new MockFileSystem(GetMockFiles1(), MockCurrentDirectory));
```
Comment (mszymeshop) [inline src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs:36]: You should mock `SubscribersDirectoryProcessor` for the purpose of Command tests

Comment (mszymeshop) [inline src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs:36]: GetMockFiles1?

File: src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+            _configureEdaCommand = new ConfigureEdaCommand(subscribersDirectoryProcessor, _apiConsumer);
+        }
+
+        [Fact, IsUnit]
+        public async Task OnExecuteAsync_WhenSingleFileRequestAccepted_Returns0()
+        {
+            // Arrange
+            _configureEdaCommand.InputFolderPath = MockCurrentDirectory;
```
Comment (mszymeshop) [inline src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs:44]: looks like this is your default anyway, consider moving to setup

File: src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs [hunk 5]
```diff
@@ -196,15 +200,34 @@ public async Task OnExecuteAsync_WhenNoDryRunFalse_ApiIsNotCalled()
...
+                .Contain($@"Reading files from folder: '{MockCurrentDirectory}' to be run against CI environment")
                 .And.Contain("File 'sample1.json' has been found")
                 .And.NotContain("Starting to run configuration against Captain Hook API")
                 .And.Contain("By default the CLI runs in 'dry-run' mode. If you want to run the configuration against Captain Hook API use the '--no-dry-run' switch")
                 .And.Contain("Processing finished");
         }
 
-        public static Dictionary<string, MockFileData> GetMockFiles1()
+        [Fact, IsUnit]
+        public async Task OnExecuteAsync_WhenInvalidJson_Returns1()
```
Comment (mszymeshop) [inline src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs:211]: Some tests contain a digit at the end of the method name

Comment (nikhilgirraj) [reply src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs:211]: The digit at the end of the name suggest the return value from the method under test. Would you rather name it differently?

Comment (nikhilgirraj) [reply src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs:211]: The digit at the end of the name suggest the return value from the method under test. How would you rather name it then?

File: src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs (hunk 3 above)
Comment (nikhilgirraj) [reply src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs:36]: It returns a dictionary of 1 mock file. Will change the approach to naming these

Comment (nikhilgirraj) [reply src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs:36]: Next task. :)

File: src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs (hunk 2 above)
Comment (nikhilgirraj) [reply src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs:34]: Next task

File: src/Platform.Eda.Cli/Commands/ConfigureEda/ConfigureEdaCommand.cs (hunk 1 above)
Comment (nikhilgirraj) [reply src/Platform.Eda.Cli/Commands/ConfigureEda/ConfigureEdaCommand.cs:54]: Yes, apparently there is a `console.EmitWarning` method called on it.

File: src/Platform.Eda.Cli/Commands/ConfigureEda/ApiConsumer.cs [hunk 6]
```diff
@@ -13,7 +13,7 @@
 
 namespace Platform.Eda.Cli.Commands.ConfigureEda
 {
-    public class ApiConsumer
+    public class ApiConsumer : IApiConsumer
```
Comment (chomick) [inline src/Platform.Eda.Cli/Commands/ConfigureEda/ApiConsumer.cs:16]: Small one - maybe 'CaptainHookApiProxy' would be more appropriate name?

File: src/Platform.Eda.Cli/Commands/ConfigureEda/ConfigureEdaCommand.cs (hunk 1 above)
Comment (mszymeshop) [reply src/Platform.Eda.Cli/Commands/ConfigureEda/ConfigureEdaCommand.cs]: Happy to use WriteError however up to you

File: src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs (hunk 5 above)
Comment (mszymeshop) [reply src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs:209]: For some reason I was surprised with the digits :) I guess your explanation makes sense. Do you think names WithOneFile, WithTwoFiles make more sense?

**Review by mszymeshop — COMMENTED** [review]
Let's fix the easy bits and merge. The deeper changes we can address as a separate task.

File: src/Platform.Eda.Cli/Commands/ConfigureEda/ConfigureEdaCommand.cs (hunk 1 above)
Comment (nikhilgirraj) [reply src/Platform.Eda.Cli/Commands/ConfigureEda/ConfigureEdaCommand.cs]: done already :)

File: src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs (hunk 5 above)
Comment (nikhilgirraj) [reply src/Tests/Platform.Eda.Cli.Tests/Commands/ConfigureEda/ConfigureEdaCommandTests.cs:209]: Had changed the methods for returning mock files to GetSingleMockInputFile, GetMultipleMockInputFiles and GetInvalidJsonMockInputFile.

File: src/Platform.Eda.Cli/Commands/ConfigureEda/ApiConsumer.cs (hunk 6 above)
Comment (nikhilgirraj) [reply src/Platform.Eda.Cli/Commands/ConfigureEda/ApiConsumer.cs:16]: Makes sense. :)

Comment (nikhilgirraj) [reply src/Platform.Eda.Cli/Commands/ConfigureEda/ApiConsumer.cs:16]: I looked at the code and thought that it's not really a CaptainHookApiProxy, much rather a CaptainHookPutSubscriberApiProxy. Would `CreateSubscriberApiProxy` be a better name?

Comment (mszymeshop) [reply src/Platform.Eda.Cli/Commands/ConfigureEda/ApiConsumer.cs:16]: Should be a noun in the end, so maybe `CaptainHookSuscriberApiProxy` :D

File: src/Platform.Eda.Cli/Commands/ConfigureEda/IApiConsumer.cs [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+using System.Linq;
+using System.Text;
+using CaptainHook.Api.Client;
+using Platform.Eda.Cli.Commands.ConfigureEda.Models;
+
+namespace Platform.Eda.Cli.Commands.ConfigureEda
+{
+    public interface IApiConsumer
+    {
+        public IAsyncEnumerable<ApiOperationResult> CallApiAsync(IEnumerable<PutSubscriberFile> files);
```
Comment (mszymeshop) [inline src/Platform.Eda.Cli/Commands/ConfigureEda/IApiConsumer.cs:12]: is the public in interface excessive? It is public by design.

File: src/Platform.Eda.Cli/Program.cs [hunk 8]
```diff
@@ -80,7 +81,14 @@ private static IServiceProvider SetupAutofac()
 
             var builder = new ContainerBuilder();
             builder.RegisterAssemblyModules(Assembly.GetExecutingAssembly());
+
+            builder.RegisterType<SubscribersDirectoryProcessor>().As<ISubscribersDirectoryProcessor>();
+            builder.RegisterType<ApiConsumer>().As<IApiConsumer>();
```
Comment (mszymeshop) [inline src/Platform.Eda.Cli/Program.cs:86]: That one you don't need as the delegate has been registered by my code already.

**Review by mszymeshop — APPROVED** [review]

File: src/Platform.Eda.Cli/Program.cs (hunk 8 above)
Comment (nikhilgirraj) [reply src/Platform.Eda.Cli/Program.cs]: I've not included that in my latest commit then.

File: src/Platform.Eda.Cli/Commands/ConfigureEda/IApiConsumer.cs (hunk 7 above)
Comment (nikhilgirraj) [reply src/Platform.Eda.Cli/Commands/ConfigureEda/IApiConsumer.cs]: Yes. Result of creating an interface from existing class. Removed.

**Pull request merged**